text/0000-macro-at-most-once-rep.md

```diff
@@ -0,0 +1,174 @@
+- Feature Name: macro-at-most-once-rep
+- Start Date: 2018-01-17
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+
+Summary
+-------
+
+Add a repetition specifier to macros to repeat a pattern at most once: `$(pat)?`. Here, `?` behaves like `+` or `*` but represents at most one repetition of `pat`.
+
+Motivation
+----------
+
+There are two specific use cases in mind.
+
+## Macro rules with optional parts
+
+Currently, you just have to write two rules and possibly have one "desugar" to the other.
+
+```rust
+macro_rules! foo {
+  (do $b:block) => {
+    $b
+  }
+  (do $b1:block and $b2:block) => {
+    foo!($b1)
+    $b2
+  }
+}
+```
+
+Under this RFC, one would simply write:
+
+```rust
+macro_rules! foo {
+  (do $b1:block $(and $b2:block)?) => {
+    $b1
+    $($b2)?
+  }
+}
+```
+
+## Trailing commas
+
+Currently, the best way to make a rule tolerate trailing commas is to create another identical rule that has a comma at the end:
+
+```rust
+macro_rules! foo {
+  ($(pat),*,) => { foo!( $(pat),* ) };
+  ($(pat),*) => {
+    // do stuff
+  }
+}
+```
+
+or to allow multiple trailing commas:
+
+```rust
+macro_rules! foo {
+  ($(pat),* $(,)*) => {
+    // do stuff
+  }
+}
+```
+
+This is unergonomic and clutters up macro definitions needlessly. Under this RFC, one would simply write:
+
+```rust
+macro_rules! foo {
+  ($(pat),* $(,)?) => {
```

Comment:
It sounds like better support for trailing delimiters is a major part of the motivation for this RFC and why people are excited about it. But the old approach shown under "Currently" still seems better than the one here because with `?` we cannot disallow an invocation consisting of only a comma, `foo!(,)`. Is the idea that `$(,)?` would be a quick and easy shorthand but macro authors would continue to use the old approach if they want their macro to be maximally correct? If so, could you brainstorm some ideas for how optional trailing delimiters could be better supported?

Reply:
Hmm... that's true. It works well if you have + repetition.

One option is to combine ? with | as described in https://github.com/rust-lang/rfcs/pull/2298#issuecomment-364681597

Reply:
Another option is to create another repetition mode which indicates trailing commas: `$[pat],*`